File: R/Lesson_012_design_add_and_search_words_data_structure.R
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 012 -- Design Add And Search Words Data Structure
# Category   : Tries
# Difficulty : Medium
# Study Plan : Day 6
# =============================================================
#
# QUESTION:
#   Design a data structure that supports:
#     addWord(word)
#     search(word)  - word may contain '.' which matches any single letter
#
#   Example:
#     d = WordDictionary(); d.addWord("bad"); d.addWord("dad"); d.addWord("mad")
#     d.search("pad") -> False
#     d.search("bad") -> True
#     d.search(".ad") -> True
#     d.search("b..") -> True
# =============================================================

new_wd <- function() { e <- new.env(hash=TRUE); e$end <- FALSE; e$ch <- new.env(hash=TRUE); e }
addWord <- function(d, w) {
    n <- d
    for (c in strsplit(w, "")[[1]]) {
        if (!exists(c, envir=n$ch, inherits=FALSE)) assign(c, new_wd(), envir=n$ch)
        n <- get(c, envir=n$ch)
    }
    n$end <- TRUE
}
search_wd <- function(d, w) {
    chars <- strsplit(w, "")[[1]]
    dfs <- function(i, n) {
        if (i > length(chars)) return(n$end)
        c <- chars[i]
        if (c == ".") {
            for (k in ls(n$ch)) if (dfs(i+1, get(k, envir=n$ch))) return(TRUE)
            return(FALSE)
        }
        if (!exists(c, envir=n$ch, inherits=FALSE)) return(FALSE)
        dfs(i+1, get(c, envir=n$ch))
    }
    dfs(1, d)
}

d <- new_wd(); for (w in c("bad","dad","mad")) addWord(d, w)
print(c(search_wd(d,"pad"), search_wd(d,"bad"), search_wd(d,".ad"), search_wd(d,"b..")))

File: Go/Lesson_012_design_add_and_search_words_data_structure.go
//go:build ignore

// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 012 -- Design Add And Search Words Data Structure
// Category   : Tries
// Difficulty : Medium
// Study Plan : Day 6
// =============================================================
//
// QUESTION:
//   Design a data structure that supports:
//     addWord(word)
//     search(word)  - word may contain '.' which matches any single letter
//
//   Example:
//     d = WordDictionary(); d.addWord("bad"); d.addWord("dad"); d.addWord("mad")
//     d.search("pad") -> False
//     d.search("bad") -> True
//     d.search(".ad") -> True
//     d.search("b..") -> True
// =============================================================

package main

import "fmt"

type WordDictionary struct { ch map[byte]*WordDictionary; end bool }
func NewWD() *WordDictionary { return &WordDictionary{ch: map[byte]*WordDictionary{}} }
func (d *WordDictionary) AddWord(w string) {
    n := d
    for i := 0; i < len(w); i++ {
        c := w[i]
        if _, ok := n.ch[c]; !ok { n.ch[c] = NewWD() }
        n = n.ch[c]
    }
    n.end = true
}
func (d *WordDictionary) Search(w string) bool { return dfs(0, w, d) }
func dfs(i int, w string, n *WordDictionary) bool {
    if i == len(w) { return n.end }
    c := w[i]
    if c == '.' {
        for _, x := range n.ch { if dfs(i+1, w, x) { return true } }
        return false
    }
    nx, ok := n.ch[c]
    return ok && dfs(i+1, w, nx)
}

func main() {
    d := NewWD()
    for _, w := range []string{"bad","dad","mad"} { d.AddWord(w) }
    fmt.Println(d.Search("pad"), d.Search("bad"), d.Search(".ad"), d.Search("b.."))
}
